op.c

```diff
@@ -16564,71 +16634,70 @@ Perl_subsignature_append_positional(pTHX_ PADOFFSET padix, OPCODE defmode, OP *d
+        yyerror("Mandatory parameter follows optional parameter");
+        return;
+    }
+
+    UV argix = signature->next_argix;
+    signature->next_argix++;
+
+    if(!padix && !defexpr)
+        /* This param has no var and no defaulting expression. There's 
+         * nothing else for us to do here.
+         */
+        return;
+
+    struct yy_parser_signature_param *param = subsignature_push_param();
+    param->argix = argix;
 
-    OP *varop = NULL;
     if(padix) {
         assert(PadnamePV(PadnamelistARRAY(PL_comppad_name)[padix])[0] == '$');
 
-        varop = newUNOP_AUX(OP_ARGELEM, 0, NULL, NULL);
-        varop->op_private |= OPpARGELEM_SV;
-        varop->op_targ = padix;
-        cUNOP_AUXx(varop)->op_aux = INT2PTR(UNOP_AUX_item *, argix);
-    }
+        param->padix = padix;
 
-    signature->elems++;
+        intro_my(); /* introduce the new pad variable now */
+    }
+    else
+        param->padix = 0;
 
     if(defexpr) {
-        signature->optelems++;
+        signature->opt_params++;
 
         if(defexpr->op_type == OP_NULL && !(defexpr->op_flags & OPf_KIDS))
         {
             /* caller passed in newOP(OP_NULL, 0), so we should not leak it */
             op_free(defexpr);
 
+            defexpr = newOP(OP_STUB, OPf_WANT_SCALAR);
+
             /* handle '$=' special case */
-            if(varop)
+            if(padix)
                 yyerror("Optional parameter lacks default expression");
         }
-        else {
-            I32 flags = 0;
-            if(defmode == OP_DORASSIGN)
-                flags |= OPpARG_IF_UNDEF << 8;
-            if(defmode == OP_ORASSIGN)
-                flags |= OPpARG_IF_FALSE << 8;
-
-            /* a normal '=default' expression */
-            OP *defop = newARGDEFELEMOP(flags, defexpr, argix);
-
-            if(varop) {
-                varop->op_flags |= OPf_STACKED;
-                (void)op_sibling_splice(varop, NULL, 0, defop);
-                scalar(defop);
-            }
-            else
-                varop = newUNOP(OP_NULL, 0, defop);
 
-            LINKLIST(varop);
-            /* NB: normally the first child of a logop is executed before the
-             * logop, and it pushes a boolean result ready for the logop. For
-             * ARGDEFELEM, the op itself does the boolean calculation, so set
-             * the first op to it instead.
+        /* generate the COP here so the line number is correct */
+        param->defcop = newSTATEOP(0, NULL, NULL);
+
+        param->defmode = defmode;
+        param->defexpr = defexpr;
+        scalar(defexpr);
+
+        if(!padix) {
+            /* A defaulting expression with anonymous var still needs to be
+             * conditional. We'll just have to allocate a pad temporary to
+             * store the value in anyway, so OP_PARAMTEST can check it.
              */
-            varop->op_next   = defop;
-            defexpr->op_next = varop;
+            param->padix = allocmy("", 0, 0);
```

Comment:
`Perl_allocmy` currently does not support empty names because it blindly uses `name[1]`. It needs to be changed to check `len` first.

Reply:
Adding more asserts in https://github.com/Perl/perl5/pull/23583 which would have caught this